dot/state/roottottrie.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/trie"
+)
+
+type rootToTrieMap struct {
+	rootToTrie map[common.Hash]*trie.Trie
+	sync.RWMutex
+}
+
+func newRootToTrieMap() *rootToTrieMap {
+	return &rootToTrieMap{
+		rootToTrie: make(map[common.Hash]*trie.Trie),
+	}
+}
+
+func (r *rootToTrieMap) get(root common.Hash) (t *trie.Trie, has bool) {
+	r.RLock()
+	defer r.RUnlock()
+	t, has = r.rootToTrie[root]
+	return t, has
+}
+
+func (r *rootToTrieMap) set(root common.Hash, trie *trie.Trie) {
+	r.Lock()
+	defer r.Unlock()
+	r.rootToTrie[root] = trie
+}
+
+func (r *rootToTrieMap) setIfUnset(root common.Hash, trie *trie.Trie) {
+	r.Lock()
+	defer r.Unlock()
+
+	_, exists := r.rootToTrie[root]
+	if exists {
+		return
+	}
+
+	r.rootToTrie[root] = trie
+}
+
+func (r *rootToTrieMap) delete(root common.Hash) {
+	r.Lock()
+	defer r.Unlock()
+	delete(r.rootToTrie, root)
+}
+
+func (r *rootToTrieMap) getFromMemOrDB(root common.Hash, db chaindb.Database) (
```

Comment:
can we move this code into the caller of this function `StorageState.TrieState()`. Doesn't make a lot of sense for the in-memory `rootToTrieMap` has to worry about offline storage `chaindb.Database`.

Reply:
>> Doesn't make a lot of sense...

Actually it does, I think it should be an atomic operation to get from memory + get from db + set to memory.

Otherwise we could end up with cases where

- goroutine A gets root hash XYZ from memory, it's not found
- goroutine B sets trie at root hash XYZ in memory
- goroutine A gets stored trie for root hash XYZ from disk
- goroutine A sets trie for root hash XYZ in memory

Effectively rendering the write of B erased. Probably won't happen, but better safe 😉

Reply:
Then shouldn't `chaindb` be stored in `rootToTrieMap` instead of being passed on every call to `getFromMemOrDB`.  Also after looking at the calling code in `StorageState` it looks it already checks the map if it has the key, and will load from db in `StorageState.loadTrie`, and it only needs this atomic read in `StorageState.TrieState`.  If it were me, I'd simplify this and just add `Lock()`, `Unlock()` methods to the `rootToTrieMap`.

Reply:
Changed pretty much a lot of code in there; now more operations are atomic and the newer `tries` structure handles the operations of memory + database.

Reply:
Moved changes regarding atomic operations in #2250